librarian/data/meta/archive.py

```diff
@@ -0,0 +1,673 @@
+        self._tasks = kwargs.get('tasks', exts.tasks)
+        self._events = kwargs.get('events', exts.events)
+        self._events.subscribe(self.ENTRY_POINT_FOUND, self._entry_point_found)
+        # Select what from meta table
+        self._meta_what = ','.join('{}.{}'.format(self.META_TABLE, c)
+                                   for c in self.META_COLUMNS)
+
+    def _analyze(self, path, partial, callback):
+        """
+        Return(and optionally ``callback``) with found metadata on ``path``.
+
+        Called by the public py:meth:`~Archive.analyze` method and performs
+        the heavy lifting to obtain and return metadata, with optionally
+        invoking a ``callback`` function with obtained metadata, if it was
+        specified.
+        """
+        logging.debug(u"Analyze[%s] %s", ('FULL', 'PARTIAL')[partial], path)
+        data = dict()
+        for proc_cls in self.Processor.for_path(path):
+            proc = proc_cls(path, data=data, partial=partial, fsal=self._fsal)
+            # store entry point on parent folder if available
+            if proc_cls.is_entry_point(path):
+                content_type = self.ContentTypes.to_bitmask(proc_cls.name)
+                self._events.publish(self.ENTRY_POINT_FOUND,
+                                     path=path,
+                                     content_type=content_type)
+            # gather metadata from current processor into ``data``
+            proc.process()
+        # invoke specified ``callback`` if available with gathered metadata
+        # and then return the same
+        wrapped = dict((k, self.MetaWrapper(v)) for (k, v) in data.items())
+        if callback:
+            callback(wrapped)
+        return wrapped
+
+    @as_iterable(params=[1])
+    @batched(arg=1, batch_size=100, aggregator=batched.updater)
+    def analyze(self, paths, partial=False, callback=None):
+        """
+        Analyze ``paths`` to determine their content type and metadata.
+
+        A comprehensive analysis will be performed only if ``partial`` is not
+        set, which also write the found metadata into the database. Otherwise,
+        if partial is set, it will perform a quick analysis and return only
+        basic information about the paths. The optional ``callback`` argument
+        determines if the analysis will run asynchronously, invoking the
+        ``callback`` function with the obtained data, or in blocking mode,
+        returning the data.
+        """
+        if not callback:
+            ret_val = dict()
+            for path in paths:
+                ret_val.update(self._analyze(path, partial, callback))
+            return ret_val
+        # schedule background task to perform analysis of ``paths``
+        func = functools.partial(self._analyze,
+                                 partial=partial,
+                                 callback=callback)
+        self._tasks.schedule(lambda paths: map(func, paths), args=(paths,))
```

Comment:
That is different from the sync version, which returns the whole batch at once. Can we try doing the same for the async one as well, unless you remember the reason ?